P0009r8: Math-ify mdspan.subspan; clean up accessor policy; fix #71

mhoemmen opened the pull request:
- Math-ify mdspan.subspan, as discussed with @crtrott today in #74 comments
- Clean up accessor policy, per discussion with @crtrott today
- Fix #71, at least partly

"At least partly" means that we may still need to look at the definitions of `required_span_size()` for specific layouts.

crtrott commented:
Looks pretty good to me: One thing: do we need the `only` in `a handle to a contiguous set of elements of type element_type, accessible only through the policy’s access function;`?

Also does the `other than slices...,` serves a particular purposes in `If i... is a pack of integers other than slices..., then denote the value of the k -th member of i... by i k.` and do we need the same sentence for `j...`?

crtrott commented:
And thanks for all your hard work, in particular since I made you redo a bit of it .... !!

mhoemmen commented:
@crtrott wrote:
> And thanks for all your hard work, in particular since I made you redo a bit of it .... !!

Awww, don't worry, it wasn't so bad :-) Thanks for taking a look tonight!  I'll do the following:

  1. Get rid of "only" in the sentence you mentioned
  2. Change the "other than `slices...`" bit: "Denote the value of the k-th member of `i...` by `<math>i<sub>k</sub></math>`, and denote the value of k-th member of `j...` by <math>j<sub>k</sub></math>`."

mhoemmen commented:
@crtrott btw there are some issues that I opened, that I think still need to be addressed: #75, #76, and #78.  I favor separating those fixes into separate pull requests.

mhoemmen commented:
@crtrott I just pushed a commit to this PR addressing your suggestions above, along with a `<math>...</math>` syntax error and some minor formatting issues.

dhollman commented on P0009/P0009.bs:
```diff
@@ -607,36 +607,42 @@ Description
+`constexpr` operation that is trivially inlined and optimized.
+Compiler vendors may apply optimizations such as loop-invariant code
+motion, including partial evaluation of multi-index layout mappings
+when indices are loop invariant.
```
(minor note) "loop-invariant" was correct in the first place, since invariant is an adjective rather than a verb here, and "loop" doesn't have the traditional form of an adverb.  (Just a grammar nit-pick :-P)

mhoemmen replied:
Fixed in the comments I just pushed.

dhollman commented on P0009/P0009.bs:
```diff
@@ -781,10 +813,10 @@ Y8b.     .d8""8b. Y88b. Y8b.     888  888 Y88b.       X88
 
-1. An `extents` object defines a *multidimensional index space* which is the Cartesian product of integers extents `[0..N0) * [0..N1) *`....
-2. The *dynamic extents* of an `extents` object correspond to the `StaticExtents` template parameters that are equal to `dynamic_extent`.  Let *DynamicRank[i]* denote the index of the *i*th such extent in the `StaticExtents` template parameter pack, and let *DynamicIndex[r]* indicate the number of such extents in the first *r* entries of the `StaticExtents` parameter pack
+1. An `extents` object defines a multidimensional index space.
```
We've included the definition of "multidimensional index space" in the introduction but not in the wording. Doesn't it also need to be in the wording?

dhollman commented on P0009/P0009.bs:
```diff
@@ -978,26 +1034,35 @@ template<ptrdiff_t... LHS, ptrdiff_t... RHS>
+      Denote the <math>r</math>-th member of packs `i...` and `j...` by 
+      <math>i<sub>r</sub></math> and <math>j<sub>r</sub></math>, respectively. 
+      Let `r` equal <math>r</math>.
+      `sizeof...(i)==E::rank()` is `true`, 
```
Not sure why we say "is `true`" here, but if that's an LWG thing it's fine.

mhoemmen replied:
@dhollman This is an LWG thing.  Boolean code-font statements are just statements that could evaluate to `false` or `true`.

mhoemmen replied:
It's different when something Returns a `bool`; then it just returns the value of the code-font `bool` statement.

dhollman commented on P0009/P0009.bs:
```diff
@@ -1018,56 +1083,63 @@ Table � — Layout mapping policy and layout mapping requirements
+       if all members of `j...` and `i...` are equal
+       except for exactly one r-th member where
+       <math>j<sub>r</sub></math> equals <math>i<sub>r</sub></math> + 1, 
+       then `m(j...) - m(i...)` equals <math>&sigma;<sub>r</sub></math>.</td>
```
Does this also need to say `sigma_r > 1`?

mhoemmen replied:
`layout_left``mdspan` are strided (`is_strided()` is `true`).  Consider the left-most index.  There, `m(j...) - m(i...)` equals 1.  Thus, <math>sigma<sub>r</sub></math> may be 1 in this case.

Whether <math>sigma<sub>r</sub></math> may be zero relates to the issue of degenerate mdspan, #78.

dhollman commented on P0009/P0009.bs:
```diff
@@ -1207,8 +1279,8 @@ constexpr mapping(const mapping<OtherExtents>& other);
 * *Effects:* Initializes `extents_` with `other.extents()`.
-* *Postconditions:* `extents()==other.extents_`.
-* *Throws:* nothing.
+* *Ensures:* `extents()==other.extents_` is `true`.
```
should this be `other.extents()` for consistency with the copy constructor?

dhollman commented on P0009/P0009.bs:
```diff
@@ -1618,93 +1681,131 @@ struct layout_stride {
-* Throws: nothing.
+* *Requires:* `other.extents()` meets the requirements for use in the initialization of `extents_`.
+* *Effects:* Initializes `extents_` with `other.extents()`, and initializes `strides_` with `other.strides()`.
+* *Ensures:* `extents()==other.extents() && strides()==other.strides()` is `true`.
```
Should this be two lines for consistency?

dhollman commented on P0009/P0009.bs:
```diff
@@ -1807,12 +1913,12 @@ Table �: Accessor policy requirements
   <td>`a.offset(p,i)`</td>
   <td>`A::offset_policy::pointer`</td>
-  <td>*Returns:* A pointer for the contiguous span of elements offset by `i`.  *Requires:* `A::decay(p)+i == A::offset_policy::decay(A::offset(p,i))`.</td>
+  <td>*Returns:* A pointer to the contiguous span of elements offset by `i`.  *Requires:* `A::decay(p)+i` equals `A::offset_policy::decay(A::offset(p,i))`.</td>
```
Should there be a line break here?

dhollman commented on P0009/P0009.bs:
```diff
@@ -2147,20 +2250,21 @@ template<class OtherElementType, class OtherExtents, class OtherLayoutPolicy, cl
 
+<!-- NOTE is_assignable_v<T,U> means T is assignable from U -->
+* *Constraints:*
+    + `is_assignable_v<ElementType(*)[], OtherElementType(*)[]>` is `true`, <!-- lifted directly from span wording -->
```
Isn't the accessor pointer type assignment sufficient (below)? The above wording is only in `span` because it lacks the `Accessor` to constrain the pointer type, which could come from an allocator to a container that uses fancy pointers.  I don't think we have that problem.

mhoemmen replied:
@dhollman I agree; this must be a left-over from the now deleted mdspan(span) constructor.

mhoemmen commented on P0009/P0009.bs:
```diff
@@ -607,36 +607,42 @@ Description
-partial evaluation of multi-index layout mappings when indices are
-loop-invariant.
+We intend the layout mapping of a multi-index to be a constant-time
+`constexpr` operation that is trivially inlined and optimized.
```
Fixed in the comments I just pushed.

mhoemmen commented:
Thanks y'all! :D
